virtualizarr/storage/_obstore.py

```diff
@@ -0,0 +1,246 @@
+                    obs.store.GCSStore,
+                    obs.store.HTTPStore,
+                    obs.store.S3Store,
+                    obs.store.LocalStore,
+                    obs.store.MemoryStore,
+                ),
+            ):
+                raise TypeError(f"expected ObjectStore class, got {store!r}")
+        super().__init__(read_only=True)
+        self.stores = stores
+        self.xr_obj: DataArray | Dataset = xr_obj
+
+    def __str__(self) -> str:
+        return f"ManifesStore({self.xr_obj})"
+
+    def __getstate__(self) -> dict[Any, Any]:
+        state = self.__dict__.copy()
+        stores = state["stores"]
+        for k, v in stores:
+            stores[k] = pickle.dumps(v)
+        state["stores"] = stores
+        return state
+
+    def __setstate__(self, state: dict[Any, Any]) -> None:
+        stores = state["stores"]
+        for k, v in stores:
+            stores[k] = pickle.loads(v)
+        state["stores"] = stores
+        self.__dict__.update(state)
+
+    async def get(
+        self,
+        key: str,
+        prototype: BufferPrototype,
+        byte_range: ByteRequest | None = None,
+    ) -> Buffer | None:
+        # docstring inherited
+        import obstore as obs
+
+        if "zarr.json" in key:
+            return get_zarr_metadata(self.xr_obj, key)
+        manifest_index = parse_manifest_index(key)
+        # Get path, offset, and length matching this key from the ChunkManifest
+        if manifest_index.variable == "__xarray_dataarray_variable__":
+            path = self.xr_obj.data.manifest._paths[*manifest_index.indexes]
+            offset = self.xr_obj.data.manifest._offsets[*manifest_index.indexes]
+            length = self.xr_obj.data.manifest._lengths[*manifest_index.indexes]
+        else:
+            path = self.xr_obj[manifest_index.variable].data.manifest._paths[
+                *manifest_index.indexes
+            ]
+            offset = self.xr_obj[manifest_index.variable].data.manifest._offsets[
+                *manifest_index.indexes
+            ]
+            length = self.xr_obj[manifest_index.variable].data.manifest._lengths[
+                *manifest_index.indexes
+            ]
+        store_request = find_matching_store(stores=self.stores, request_key=path)
+        # Transform the input byte range to account for the chunk location in the file
+        chunk_end_exclusive = offset + length
+        byte_range = _transform_byte_range(
+            byte_range, chunk_start=offset, chunk_end_exclusive=chunk_end_exclusive
+        )
+        # Actually get the bytes
+        try:
+            bytes = await obs.get_range_async(
+                self.stores[store_request.store_id],
+                store_request.key,
+                start=byte_range.start,
+                end=byte_range.end,
+            )
+            return prototype.buffer.from_bytes(bytes)  # type: ignore[arg-type]
+        except _ALLOWED_EXCEPTIONS:
+            return None
```

Comment:
Why are any of these exceptions allowed? Don't we want to raise them?

Reply:
This is how Zarr handles empty chunks (xref https://github.com/zarr-developers/VirtualiZarr/issues/16). I'll need to put some more thought into this and return with a more complete answer later on.

Reply:
Have you read this? https://github.com/zarr-developers/VirtualiZarr/pull/427#discussion_r1943767851

Seems related.